### ISSUE TITLE
Exception 0xc0000005 on input handling (Windows)

### ISSUE THREAD

**Issue by prattmic** [go-delve/delve#2869]
Please answer the following before submitting your issue:

Note: Please include any substantial examples (debug session output,
stacktraces, etc) as linked gists.

1. What version of Delve are you using (`dlv version`)?

```
> dlv.exe version
Delve Debugger
Version: 1.8.0
Build: $Id: 6a6c9c332d5354ddf1f8a2da3cc477bd18d2be53 $
```

This binary was built with d95f26c4bd2ad9f7a92de54bc4525219104b7ed3 with `GOOS=windows go build github.com/go-delve/delve/cmd/dlv`

2. What version of Go are you using? (`go version`)?

`go version go1.17 linux/amd64`

3. What operating system and processor architecture are you using?

windows/amd64

Specifically I am using a Windows Server 2016 [gomote](https://github.com/golang/go/wiki/Gomote) over SSH (no GUI).

4. What did you do?

I was debugging a modified version of [runtime/testdata/testprogcgo](https://cs.opensource.google/go/go/+/master:src/runtime/testdata/testprogcgo/) with a modified version of the Go runtime. N.B. delve itself was built with unmodified Go 1.17, not this hacked up runtime. I can push my code if you think it is relevant, but based on the failure I don't think so.

I started with `dlv exec`:

```
> dlv.exe exec -- go\bin\testprogcgo.exe WindowsGoCallback
Type 'help' for list of commands.
(dlv) c
new extra m 0xc00003a000
new extra m 0xc0002a0000
profilem 0x123d3e0
profilem 0xc00003a000
> runtime.profilem() c:/workdir/go/src/runtime/os_windows.go:1237 (PC: 0x10b6801)
Warning: debugging optimized function
  1232:
  1233:         sigprof(c.ip(), c.sp(), c.lr(), gp, mp)
  1234:
  1235:         if mp.thread == 0 {
  1236:                 Breakpoint()
=>1237:                 d := dlog()
  1238:                 if d != nil {
  1239:                         d.s("crashing on")
  1240:                         d.p(mp)
  1241:                         d.end()
  1242:                 }
(dlv)
```

From this point I poked around a bit without issue, primarily looking at various threads (which may have been running C code and/or created by C).

```
(dlv) tr 828
Switched from 668 to 828
(dlv) bt
0  0x00007ff9ccf65b54 in ???
   at ?:-1
1  0x00000000010b6225 in runtime.stdcall2
   at :0
2  0x00000000010b52ef in runtime.semasleep
   at c:/workdir/go/src/runtime/os_windows.go:819
3  0x000000000108e3e5 in runtime.notesleep
   at c:/workdir/go/src/runtime/lock_sema.go:181
4  0x00000000010be9ca in runtime.mPark
   at c:/workdir/go/src/runtime/proc.go:1447
5  0x00000000010bfed8 in runtime.stopm
   at c:/workdir/go/src/runtime/proc.go:2444
6  0x00000000010c0865 in runtime.startlockedm
   at c:/workdir/go/src/runtime/proc.go:2667
7  0x00000000010c2133 in runtime.schedule
   at c:/workdir/go/src/runtime/proc.go:3438
8  0x00000000010c284d in runtime.park_m
   at c:/workdir/go/src/runtime/proc.go:3552
9  0x00000000010e45ea in runtime.mcall
   at c:/workdir/go/src/runtime/asm_amd64.s:425
(dlv) tException 0xc0000005 0x1 0xc001293064 0x7ff9c966d783
PC=0x7ff9c966d783

runtime.cgocall(0x887880, 0xc000057060)
        /usr/lib/google-golang/src/runtime/cgocall.go:156 +0x4a fp=0xc00126ae80 sp=0xc00126ae48 pc=0x8246ca
syscall.Syscall6(0x7ff9cc6d4ad0, 0x4, 0x88, 0xc00129307c, 0x1, 0xc001293064, 0x0, 0x0)
        /usr/lib/google-golang/src/runtime/syscall_windows.go:493 +0xfa fp=0xc00126aeb8 sp=0xc00126ae80 pc=0x88263a
syscall.Syscall6(0xc001292f88, 0x100, 0x83e235, 0x9, 0xc00126af88, 0x82d6de, 0x250e1a76cf8, 0x1000001265a09)
        <autogenerated>:1 +0x3a fp=0xc00126af20 sp=0xc00126aeb8 pc=0x88879a
syscall.(*Proc).Call(0x82e267, {0xc0007c20c0, 0xd82460, 0xc0009fc301})
        /usr/lib/google-golang/src/syscall/dll_windows.go:189 +0x1fb fp=0xc00126b010 sp=0xc00126af20 pc=0x893f3b
syscall.(*LazyProc).Call(0xc0001786f0, {0xc0007c20c0, 0x4, 0x4})
        /usr/lib/google-golang/src/syscall/dll_windows.go:340 +0x50 fp=0xc00126b040 sp=0xc00126b010 pc=0x894f90
github.com/peterh/liner.(*State).readNext(0xc000b6e840)
        /usr/local/google/home/mpratt/src/delve/vendor/github.com/peterh/liner/input_windows.go:184 +0xdd fp=0xc00126b0a0 sp=0xc
00126b040 pc=0xc7d79d
github.com/peterh/liner.(*State).PromptWithSuggestion(0xc000b6e840, {0xe443be, 0x0}, {0x0, 0x0}, 0x0)
        /usr/local/google/home/mpratt/src/delve/vendor/github.com/peterh/liner/line.go:651 +0x805 fp=0xc00126b510 sp=0xc00126b0a
0 pc=0xc82385
github.com/peterh/liner.(*State).Prompt(...)
        /usr/local/google/home/mpratt/src/delve/vendor/github.com/peterh/liner/line.go:595
github.com/go-delve/delve/pkg/terminal.(*Term).promptForInput(0xc000b6e900)
        /usr/local/google/home/mpratt/src/delve/pkg/terminal/terminal.go:356 +0x38 fp=0xc00126b568 sp=0xc00126b510 pc=0xcc0098
github.com/go-delve/delve/pkg/terminal.(*Term).Run(0xc000b6e900)
        /usr/local/google/home/mpratt/src/delve/pkg/terminal/terminal.go:278 +0x6af fp=0xc00126b790 sp=0xc00126b568 pc=0xcbf18f
github.com/go-delve/delve/cmd/dlv/cmds.connect({0xe431d3, 0x0}, {0xf27ee8, 0xc0000ccc80}, 0x0, 0x0)
        /usr/local/google/home/mpratt/src/delve/cmd/dlv/cmds/commands.go:852 +0x1ca fp=0xc00126b828 sp=0xc00126b790 pc=0xd3086a
github.com/go-delve/delve/cmd/dlv/cmds.execute(0x0, {0xc000254660, 0x2, 0x3}, 0xc00020a1a0, {0x0, 0x0}, 0x0, {0xc000254660, 0x2,
 ...}, ...)
        /usr/local/google/home/mpratt/src/delve/cmd/dlv/cmds/commands.go:1002 +0xf7d fp=0xc00126bc08 sp=0xc00126b828 pc=0xd3189d

github.com/go-delve/delve/cmd/dlv/cmds.New.func4(0xc00024c500, {0xc000254660, 0x3, 0x3})
        /usr/local/google/home/mpratt/src/delve/cmd/dlv/cmds/commands.go:252 +0x59 fp=0xc00126bc80 sp=0xc00126bc08 pc=0xd320f9
github.com/spf13/cobra.(*Command).execute(0xc00024c500, {0xc000254630, 0x3, 0x3})
        /usr/local/google/home/mpratt/src/delve/vendor/github.com/spf13/cobra/command.go:856 +0x5f8 fp=0xc00126bd40 sp=0xc00126b
c80 pc=0xd222f8
github.com/spf13/cobra.(*Command).ExecuteC(0xc00016f680)
        /usr/local/google/home/mpratt/src/delve/vendor/github.com/spf13/cobra/command.go:960 +0x3ad fp=0xc00126bdf8 sp=0xc00126b
d40 pc=0xd2296d
github.com/spf13/cobra.(*Command).Execute(...)
        /usr/local/google/home/mpratt/src/delve/vendor/github.com/spf13/cobra/command.go:897
main.main()
        /usr/local/google/home/mpratt/src/delve/cmd/dlv/main.go:24 +0x18a fp=0xc00126bf80 sp=0xc00126bdf8 pc=0xd3276a
runtime.main()
        /usr/lib/google-golang/src/runtime/proc.go:255 +0x217 fp=0xc00126bfe0 sp=0xc00126bf80 pc=0x85a397
runtime.goexit()
        /usr/lib/google-golang/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc00126bfe8 sp=0xc00126bfe0 pc=0x886001

goroutine 20 [chan receive, 4 minutes, locked to thread]:
github.com/go-delve/delve/pkg/proc/native.(*nativeProcess).handlePtraceFuncs(0xc000134480)
        /usr/local/google/home/mpratt/src/delve/pkg/proc/native/proc.go:301 +0x77
created by github.com/go-delve/delve/pkg/proc/native.newProcess
        /usr/local/google/home/mpratt/src/delve/pkg/proc/native/proc.go:64 +0x1db

goroutine 6 [select, 2 minutes]:
net.(*pipe).read(0xc0000ccc80, {0xc000aec001, 0x1fdff, 0x85a840})
        /usr/lib/google-golang/src/net/pipe.go:159 +0x15b
net.(*pipe).Read(0xa5a, {0xc000aec001, 0xc00004c000, 0xc000cf5e18})
        /usr/lib/google-golang/src/net/pipe.go:142 +0x25
encoding/json.(*Decoder).refill(0xc00006c640)
        /usr/lib/google-golang/src/encoding/json/stream.go:165 +0x17f
encoding/json.(*Decoder).readValue(0xc00006c640)
        /usr/lib/google-golang/src/encoding/json/stream.go:140 +0xbb
encoding/json.(*Decoder).Decode(0xc00006c640, {0xd9c320, 0xc000b9db48})
        /usr/lib/google-golang/src/encoding/json/stream.go:63 +0x78
net/rpc/jsonrpc.(*clientCodec).ReadResponseHeader(0xc000b9db00, 0xc0009a4000)
        /usr/lib/google-golang/src/net/rpc/jsonrpc/client.go:76 +0x85
net/rpc.(*Client).input(0xc000b95e60)
        /usr/lib/google-golang/src/net/rpc/client.go:109 +0xba
created by net/rpc.NewClientWithCodec
        /usr/lib/google-golang/src/net/rpc/client.go:206 +0xc7

goroutine 7 [select, 2 minutes]:
net.(*pipe).read(0xc0000ccc00, {0xc000984000, 0x1000, 0xc000cf1b88})
        /usr/lib/google-golang/src/net/pipe.go:159 +0x15b
net.(*pipe).Read(0xc000f0d4a0, {0xc000984000, 0x832b14, 0xc000cf1b88})
        /usr/lib/google-golang/src/net/pipe.go:142 +0x25
bufio.(*Reader).Read(0xc000982000, {0xc0009aa001, 0x5ff, 0xc000cf1c18})
        /usr/lib/google-golang/src/bufio/bufio.go:227 +0x1b4
encoding/json.(*Decoder).refill(0xc00006c780)
        /usr/lib/google-golang/src/encoding/json/stream.go:165 +0x17f
encoding/json.(*Decoder).readValue(0xc00006c780)
        /usr/lib/google-golang/src/encoding/json/stream.go:140 +0xbb
encoding/json.(*Decoder).Decode(0xc00006c780, {0xd9c380, 0xc000b95fa0})
        /usr/lib/google-golang/src/encoding/json/stream.go:63 +0x78
net/rpc/jsonrpc.(*serverCodec).ReadRequestHeader(0xc000b95f80, 0xc000f15ec0)
        /usr/lib/google-golang/src/net/rpc/jsonrpc/server.go:66 +0x85
github.com/go-delve/delve/service/rpccommon.(*ServerImpl).serveJSONCodec(0xc0000ded20, {0xf16188, 0xc000986000})
        /usr/local/google/home/mpratt/src/delve/service/rpccommon/server.go:293 +0x29d
created by github.com/go-delve/delve/service/rpccommon.(*ServerImpl).serveConnectionDemux
        /usr/local/google/home/mpratt/src/delve/service/rpccommon/server.go:188 +0x3e5

goroutine 9 [syscall, 4 minutes]:
os/signal.signal_recv()
        /usr/lib/google-golang/src/runtime/sigqueue.go:169 +0x98
os/signal.loop()
        /usr/lib/google-golang/src/os/signal/signal_unix.go:24 +0x19
created by os/signal.Notify.func1.1
        /usr/lib/google-golang/src/os/signal/signal.go:151 +0x2c

goroutine 10 [chan receive, 4 minutes]:
github.com/go-delve/delve/pkg/terminal.(*Term).sigintGuard(0xc000b6e900, 0xc17b66, 0x0)
        /usr/local/google/home/mpratt/src/delve/pkg/terminal/terminal.go:148 +0x46
created by github.com/go-delve/delve/pkg/terminal.(*Term).Run
        /usr/local/google/home/mpratt/src/delve/pkg/terminal/terminal.go:214 +0x171
rax     0x1
rbx     0x0
rcx     0xf405790625650000
rdi     0x0
rsi     0xc001293064
rbp     0xc00126ae70
rsp     0x56bdbff580
r8      0x56bdbff418
r9      0x56bdbff520
r10     0x0
r11     0x246
r12     0xc00126afa0
r13     0x0
r14     0xc00004c000
r15     0xffffffffffffffff
rip     0x7ff9c966d783
rflags  0x10246
cs      0x33
fs      0x53
gs      0x2b

gopher@SERVER-2016-V7- c:\workdir>
```

Note that the crash occurred when I was attempting to type `tr 123` (switch to another thread). It perhaps was triggered by my press of `r`, I'm not sure. But I did not press enter and thus run a command. The stack trace indeed looks like an issue with input handling.

5. What did you expect to see?

No crash when typing input.

6. What did you see instead?

Crash. See above.

**Comment by prattmic** [comment]
Also worth mentioning is that my breakpoint is in [`profilem`](https://cs.opensource.google/go/go/+/master:src/runtime/os_windows.go;l=1212?q=profilem&ss=go) at which point the thread being profiled is suspended (with _SuspendThread, see profileloop).

My intention with this debugging session was to inspect the suspended thread and see what point it is at. Honestly, I'm not certain if this is even possible for a Windows debugger, or if it would cause issues with delve. I'm also not sure if I had switched to the suspended thread prior to the crash or not.

**Comment by aarzilli** [comment]
I won't have time to look into this closely until saturday, probably, however asyncpreempt on windows uses a similar SuspendThread/GetThreadContext combo and it is incompatible with debuggers. The reason there is that it injects a function call, I don't see this here, but maybe there's something else that goes poorly. I have an issue about this on golang/go issue tracker.

**Comment by prattmic** [comment]
Thanks, good to know.

FWIW, I've encountered this crash one more time. This time, delve seemed to hang and not accept input. Pressing ^C printed "received SIGINT, stopping process (will not forward signal)" but did not exit or restore control. While poking around and trying to find the Windows equivalent of ^Z (never did figure that out...) I eventually got the same panic.

A third time delve hung the same way, but it never crashed and I eventually had to kill my SSH session to get out.
